web/src/OverviewLogPane.tsx

```diff
@@ -320,27 +372,150 @@ export class OverviewLogComponent extends Component<OverviewLogComponentProps> {
+    for (let i = 0; i < forwardLines.length; i++) {
+      let line = forwardLines[i]
+      let rendered = this.renderLineHelper(line)
+      if (!rendered) {
+        // This must be updating the progress of something that hasn't been
+        // rendered yet.
+        this.backwardBuffer.push(line)
+      }
+    }
+
+    // Render the lines in the backward buffer next.
+    let backwardStart = Math.max(0, this.backwardBuffer.length - renderWindow)
+    let backwardLines = this.backwardBuffer.slice(backwardStart)
+    this.backwardBuffer = this.backwardBuffer.slice(0, backwardStart)
+
+    let unrendered: LogLine[] = []
+    for (let i = backwardLines.length - 1; i >= 0; i--) {
+      let line = backwardLines[i]
+      let rendered = this.renderLineHelper(line)
+      if (!rendered) {
+        unrendered.push(line)
       }
-      this.elByStoredLineIndex[line.storedLineIndex] = lineEl
+    }
 
-      lastManifestName = line.manifestName
+    // Avoid infinite loops! If we can't render anything in the backward buffer,
+    // don't requeue. Just delete the backward buffer.
+    if (unrendered.length == backwardLines.length) {
+      this.backwardBuffer.length = 0
+    } else {
+      // This must be updating the progress of something that hasn't been
+      // rendered yet, push them back into the buffer.
+      this.backwardBuffer.push(...unrendered.reverse())
     }
 
     if (this.autoscroll) {
       this.scrollCursorIntoView()
     }
+
+    if (this.forwardBuffer.length || this.backwardBuffer.length) {
+      this.renderBufferRafId = this.props.raf.requestAnimationFrame(
+        this.renderBuffer
+      )
+    }
+  }
+
+  // Helper function for rendering lines. Returns true if the line was
+  // successfully rendered.
+  //
+  // If the line has already been rendered, replace the rendered line.
+  //
+  // If it hasn't been rendered, but the next line has, put it before the next line.
+  //
+  // If it hasn't been rendered, but the previous line has, put it after the previous line.
+  //
+  // If neither the previous or next line has been rendered, then this is some
+  // sort of progress indicator for a line that's outside what we're currently
+  // rendering. Return false, and let the caller put it back in the buffer.
+  renderLineHelper(line: LogLine): boolean {
+    let entry = this.lineHashList.lookup(line)
+    if (!entry) {
+      return false
+    }
+
+    let mn = this.props.manifestName
+    let showManifestName = !mn
+    let prevManifestName = entry.prev?.line.manifestName || ""
+    let isContextChange = !!entry.prev && prevManifestName !== line.manifestName
+    let lineEl = newLineEl(entry.line, isContextChange, showManifestName)
+
+    let root = this.rootRef.current
+    let existingLineEl = entry.el
+    if (existingLineEl) {
+      root.replaceChild(lineEl, existingLineEl)
+      entry.el = lineEl
+      return true
+    }
+
+    // If the only child is the cursor, then this is the first line we're trying
```

Comment:
"first" as in... the first log line of the entire tilt run? the thing we want to put on the screen first time-wise, meaning the bottom-most line? the thing we we want to put at the very top of the screen?

Reply:
Added this comment:

    // Our rendering algorithm inserts the lines incrementally by placing
    // each line next to an existing line.
    //
    // In the base case (where no lines have been rendered yet), there's only
    // one possible choice (before the cursor).

there are cases where it won't be the bottom-most line! But it doesn't really matter, because there's only one place to insert the line.

Reply:
ooh, i thought of a way to do this that has fewer edge cases, so i deleted this bit